content/posts/2018-01-10-kotlin-null.md

```diff
@@ -0,0 +1,249 @@
+
+It is totally another story about mutiple _different_ Nullability Annotations artifacts
+from JetBrains, FindBugs, The Checker Framework, Android Support Library and others,
+but [Jesse has it covered](https://medium.com/square-corner-blog/rolling-out-nullable-42dd823fbd89).
+
+# The Unicorn
+
+## The Problem
+
+### Java
+
+At least from my personal perspective, it is extremely naive to expect that every single Java library
+available will follow the lead of Nullability Annotations usage in their code.
+Asking people to adopt these is actually asking them to change their development culture,
+habits and even entire system designs. And sometimes the benefit is too small for the effort.
+
+Yes, it will help with the documentation and IDE inspections, not only with the Kotlin usage,
+but if you have a billion of public methods in a huge framework without any particular knowledge
+if a return value or an argument can be `null`...
+The situation becomes even worse with non-library and non-open source projects with deadlines and priorities.
+
+### Habits
+
+I’ve spent something about 1.5 years writing Kotlin-only code every single workday.
+Sometimes Python, sometimes Bash, but it does not change the picture.
+
+The human mindset adapts very well. Typing Kotlin code and thinking with Kotlin concepts every day
+_will_ make you forget about simple things — like the one we are discussing here — Java does not have
+nullability metadata out of the box. You will expect that everything you touch actually has it
+(a classic [Midas mistake](https://en.wikipedia.org/wiki/Midas#Golden_Touch)),
+because your shiny Kotlin codebase has it and you have a great confidence in that.
+
+Prepare for unforeseen consequences.
+
+## The Solution
+
+### The Obvious
+
+This is a hard one since it is based on a non-automated semantics mechanism — the mind.
+It is always better to remember when you are using Java code and when you are using the Kotlin one.
+
+Ask more _strange_ questions during a code review — _Are we guarded against this Java code?_
+It may sound dumb and people are often too lazy to check but it is better to raise a question instead of doing nothing.
+
+### Contributions
+
+Let’s face it — a lot of code we use as libraries is open source.
+The entire Android Framework is here — clone it, read it, modify it.
+Yes, it is not so easy to talk people into using nullability annotations,
+especially if your main argument is “I use Kotlin, so you should...”.
+
+* Maintainers are not your personal minions, so please show at least some respect.
+* The Kotlin argument does not work all the time and it should be the last one you have up your sleeve.
+  Not so many people actually use Kotlin at this point, but there is a huge amount of Java code.
+* Nullability annotations improve documentation and IDE support for everyone who uses the Java code,
+  so it seems like a better argument.
+* Pull Requests are always welcome!
+
+### Facade the Hell of It
```

Comment:
Nitpicking, but is this really the facade pattern? Although you never explicitly said "Facade pattern" but only "Facade" I somehow have the Facade Pattern in my mind which is technically slightly different from a `interface` and concrete implementation. From a pattern point of view I think Decorator pattern is closer to what you are describing in this section.

Not sure how to "fix"this nor if it's worthwhile to change it, it is actually not a big issue ... just saying ...

Reply:
I’ll change Facade to Abstract to avoid the confusion, thanks!

When I typed this I thought mostly about common sense instead of design patterns.